src/Silex/Provider/SecurityServiceProvider.php

```diff
@@ -140,14 +142,20 @@ public function register(Container $app)
             return new UserChecker();
         };
 
-        $app['security.access_manager'] = function ($app) {
-            $votersIterator = new ServiceIterator($app, $app['security.voter_services']);
+        $app['security.access_manager'] = function ($app) use ($default_voters) {
+            $voter_services = $app['security.voter_services'];
+            if (empty(array_diff($voter_services, $default_voters))) {
```

Comment:
what we actually need to know is whether `$app['security.voters']` was extended to register some in the old way, not whether `$app['security.voter_services']` was extended.

If you extend both, we still need to register the ones registered in the old way, which your code does not.
And the default implementation of `$app['security.voters']` means that any voter registered in the new way would be available in it if needed.

Reply:
A simpler implementation could be achieved if the `Pimple\ServiceIterator` class accepts either service ID or instances. A bit similar to how Doctrine listener resolvers work http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/events.html#entity-listeners-resolver